sdk/identity/azure-identity/src/main/java/com/azure/identity/IdentityClient.java

```diff
@@ -173,15 +191,37 @@ public IdentityClient(IdentityClientOptions options) {
      */
     public Mono<AccessToken> authenticateWithUsernamePassword(String tenantId, String clientId, String[] scopes, String username, String password) {
         String authorityUrl = options.authorityHost().replaceAll("/+$", "") + "/" + tenantId;
-        ExecutorService executor = Executors.newSingleThreadExecutor();
         try {
-            PublicClientApplication application = PublicClientApplication.builder(clientId)
-                .authority(authorityUrl)
-                .executorService(executor)
-                .build();
+            PublicClientApplication.Builder applicationBuilder = PublicClientApplication.builder(clientId).authority(authorityUrl);
+            if (options.proxyOptions() != null) {
+                applicationBuilder.proxy(proxyOptionsToJavaNetProxy(options.proxyOptions()));
+            }
+            PublicClientApplication application = applicationBuilder.build();
             return Mono.fromFuture(application.acquireToken(UserNamePasswordParameters.builder(new HashSet<>(Arrays.asList(scopes)), username, password.toCharArray()).build()))
-                .map(ar -> new AccessToken(ar.accessToken(), OffsetDateTime.ofInstant(ar.expiresOnDate().toInstant(), ZoneOffset.UTC)))
-                .doFinally(s -> executor.shutdown());
+                .map(ar -> {
+                    currentAccount.set(ar.account());
+                    currentApplication.set(application);
+                    return new AccessToken(ar.accessToken(), OffsetDateTime.ofInstant(ar.expiresOnDate().toInstant(), ZoneOffset.UTC));
+                });
+        } catch (MalformedURLException e) {
+            return Mono.error(e);
+        }
+    }
+
+    public Mono<AccessToken> authenticateWithCurrentlyLoggedInAccount(String[] scopes) {
+        try {
+            SilentParameters parameters;
+            if (currentApplication.get() != null) {
+                if (currentAccount.get() != null) {
+                    parameters = SilentParameters.builder(new HashSet<>(Arrays.asList(scopes)), currentAccount.get()).build();
+                } else {
+                    parameters = SilentParameters.builder(new HashSet<>(Arrays.asList(scopes))).build();
+                }
+                return Mono.fromFuture(currentApplication.get().acquireTokenSilently(parameters))
+                    .map(ar -> new AccessToken(ar.accessToken(), OffsetDateTime.ofInstant(ar.expiresOnDate().toInstant(), ZoneOffset.UTC)));
+            } else {
+                return Mono.error(new UnsupportedOperationException("No cached token available - Must call another method to authenticate first"));
```

Comment:
Why throw an exception here vs. returning no AccessToken?

Reply:
Because all the other methods on the IdentityClient either returns a valid token or throws?

Reply:
Synced offline - returning null will not get through reactor thus throwing.